### ISSUE TITLE
Pose Estimation Ouput format

### ISSUE THREAD

**Issue by Sadat75** [ultralytics/ultralytics#7414]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello,

Currently, I get the output of key points from the Yolov8n-pose model in the format:
px1 py1, px2 py2

How can I get the visibility score as well in the format:
px1 py1 p1-visibility, px2 py2 p2-visibility

Thank you.

**Comment by glenn-jocher** [comment]
@Sadat75 hello!

The YOLOv8n-pose model outputs keypoints in the format you've described (px1 py1, px2 py2, etc.), which represent the coordinates of the detected keypoints. Currently, the visibility score is not included in the output format.

If you're looking to include visibility scores for each keypoint, this would require a modification to the model's output processing. While this feature isn't available out-of-the-box, it's a valuable suggestion for future updates. We appreciate your input and encourage you to keep an eye on our updates or consider contributing to the repo.

For now, you can refer to our documentation on pose estimation for more details on how keypoints are handled within the YOLOv8 framework. Thank you for your interest in YOLOv8 and for being part of our community! 😊

**Comment by Sadat75** [comment]
@glenn-jocher
 
Thank you for your response!

**Comment by jrajpal5-singularity** [comment]
> Hi @glenn-jocher, is the same feature available now which also predicts the visibility score at the end?

Hi @glenn-jocher , the feature would help in the auto-annotation pipeline to further improve the model accuracy.

**Comment by glenn-jocher** [comment]
Hello @jrajpal5-singularity,

As of now, the feature to predict visibility scores alongside keypoints is not yet implemented in YOLOv8. We understand how this could enhance auto-annotation pipelines and appreciate your suggestion. Keep an eye on our repository updates for any future enhancements. Thank you for your input! 😊

**Comment by jrajpal5-singularity** [comment]
Hello @glenn-jocher  Is there any possibility of this feature in near future?

**Comment by glenn-jocher** [comment]
@jrajpal5-singularity hello,

The YOLOv8n-pose model provides keypoint detection, which includes the coordinates of detected keypoints. However, by default, the visibility score is not included in the output. The visibility score typically indicates whether a keypoint is visible (occluded or not) in the image.

To obtain the visibility score along with the keypoint coordinates, you would need to modify the model's post-processing code to include this information in the output. This might involve changes to the model's inference script where the keypoints are parsed and returned.

If the model has been trained to predict visibility scores and these scores are part of the model's outputs, you can access them by modifying the `Results` object's `keypoints` attribute. Here's a general approach on how you might extract and format the keypoints with visibility scores:

```python
from ultralytics import YOLO

# Load a pretrained YOLOv8n-pose model
model = YOLO('yolov8n-pose.pt')

# Run inference on an image
results = model('path/to/image.jpg')

# Process results
for result in results:
    # Assuming keypoints are stored in a (N, 3) tensor where N is the number of keypoints
    # and the columns represent [x, y, visibility]
    keypoints_with_visibility = result.keypoints.xyv  # This attribute might not exist and needs to be implemented

    # Format the keypoints with visibility
    formatted_keypoints = []
    for kp in keypoints_with_visibility:
        x, y, visibility = kp
        formatted_keypoints.append(f"{x} {y} {visibility}")

    # Join formatted keypoints into a single string
    keypoints_str = ', '.join(formatted_keypoints)
    print(keypoints_str)
```

Please note that the `xyv` attribute (containing the x, y coordinates and visibility) used in the example above is hypothetical and would need to be implemented in the `Results` class to reflect the actual model outputs.

If the visibility scores are not currently part of the model's outputs, you would need to retrain the model with annotations that include visibility information and adjust the model architecture to output these scores.

Keep in mind that modifying the model and its outputs may require a good understanding of the model architecture and the data it was trained on. If you're not familiar with these aspects, you might need to consult with someone who has experience in model development and pose estimation tasks.

**Comment by Ashutosh1995** [comment]
Hi @glenn-jocher, can you please tell your insights as to how can we proceed for the training such a model that also gives visibility at the inference? Right now, I have a 12 GB GPU and I believe I can't train M,L models on it with batch size >4,2. I would want to work on this problem statement.

**Comment by glenn-jocher** [comment]
@Ashutosh1995 hi there!

Training a model to predict visibility scores during inference is an interesting challenge. You'll need to ensure your dataset includes visibility annotations for each keypoint. Once you have the annotated dataset, you can modify the model architecture to output an additional value per keypoint representing the visibility score.

Regarding your GPU memory constraints, you're correct that larger models require more memory. However, you can still train smaller models effectively with a batch size of 4 or even use techniques like gradient accumulation to simulate larger batch sizes. Here's a quick example of how you might adjust your training command:

```bash
yolo train data=your_dataset.yaml model=yolov8n.yaml batch=4 accumulate=2
```

This way, you can manage memory usage while effectively training your model. Good luck with your project, and keep us posted on your progress! 😊

**Comment by Ashutosh1995** [comment]
Thanks @glenn-jocher for your valuable feedback, let me work on this and get back to you.
